### ISSUE TITLE
back to MinGW ?

### ISSUE THREAD

**Issue by volth** [nix-windows/nix#20]
It might have sense to switch back to MinGW as the main compiler:

1. Ironically, it is the only modern C++ compiler whose binary distribution does not require modern Windows to run.
2. Eelco loves to write GCC-only code in upstream Nix.
3. MSVC .iso file currently used for bootstrap nixpkgs is huge.

**Comment by volth** [comment]
Of course, `stdenvMsvc2005` ... `stdenvMsvc2019` won't go

**Comment by Ericson2314** [comment]
https://github.com/NixOS/nixpkgs/pull/72366/files I hope will help with bootstrapping eventually, but yes I think the first thing we should target merging upstream is MinGW + existing autoconf build system.

**Comment by volth** [comment]
`autoconf` might be too tricky to run on Windows :)

I meant, while source form of distribution (and installing via `./configure && make install`) is common on Unix'es, given big zoo of different architectures, glibc, gccs, ..., on Windows it is easy to make an executable file (`nix.exe` to read `nixpkgs` and `fetchurl`/`fetchfromGutHub` everything) which can run on every Windows starting from ancient ones to the most modern.
I am here only about the first C++ compiler to download.

**Comment by Ericson2314** [comment]
Autoconf should be fine in MSYS though? (But using the mingw toolchain not msys one so the built nix doesn't require MSYS.)

I agree Autoconf is the wrong technology, but I'm quite anxious about trying to break up the work into upstreamable bits. If the MSYS-requiring build is a small enough diff but also motivates more abstractions to minimize CPP, that's a beachhead.

**Comment by AtnNn** [comment]
I think it's worth trying getting autotools to work. I'm a fan of the [MSYS2 approach](https://www.msys2.org/wiki/How-does-MSYS2-differ-from-Cygwin/): a Cygwin-based  "minimal shell required to run autotools and other build systems".

Another potential issue with MSVC is licensing and redistribution.

**Comment by volth** [comment]
Upstream Nix builds on MSYS just well (probably with minimal changes which I forgot, they might be found in early commits https://github.com/nix-windows/nix/commits/windows?after=5b2bc90ae0ac3b2e40d234f316432e0ef87e2261+34&branch=windows).

It is just has little value: everything what can be built for MSYS ecosystem using Nix is pre-built already, can be downloaded in binary form, unpack into `c:\nix\store\whatever` and happily run from there (see `msysPacman.*` and `mingwPacman.*` in `nixpkgs-windows`)

**Comment by volth** [comment]
> I think it's worth trying getting autotools to work. I'm a fan of the [MSYS2 approach](https://www.msys2.org/wiki/How-does-MSYS2-differ-from-Cygwin/): a Cygwin-based "minimal shell required to run autotools and other build systems".
> 
> Another potential issue with MSVC is licensing and redistribution.

Eventually we will have to leave the cozy matrix of Cygwin/MSYS/WSL and stem to the desert of the real.
For example, to create a device driver to support write-protected Nix Store and sandboxed builds.

**Comment by Ericson2314** [comment]
Well what I was thinking when I said "using the mingw toolchain not msys" was that we build-time depend on cozy MSYS, but run time do not need it. Hopefully that's enough "desert of the real" to motivate a first version of this? Absolutely I want to be able to do the rest, in the end :).

> For example, to create a device driver to support write-protected Nix Store and sandboxed builds.

I opened https://github.com/nix-windows/nix/issues/22 with thing they made for Docker on windows which might help.

**Comment by volth** [comment]
> we build-time depend on cozy MSYS, ...to motivate a first version of this?

Yes, but... once we got a working `nix.exe` (and we already got it 2 years ago), what would be the motivation to maintain this path?

Having `nix.exe` we can use it to download any toolchain (including but not limited to MSYS), and to build the next version on `nix.exe` (via `nix-build.exe -E "(import <nixpkgs>).pkgs.nix"` - this does not work yet, but trivial)

**Comment by Ericson2314** [comment]
The motivation is merging into upstream Nix! To be clear, I'm fine keeping the `_MSC_VER` preprocessor or other such things on all branches, but I don't think Eelco wants to replace the build system just yet. If I'm wrong, then by all means let's support more than MinGW.

Perhaps for some context, my first motivation for Nix on Windows are:

1.To support the development of software on the "big 3" with just one tool / package set. So a Nix fork that doesn't work but is never merged upstream erases the value proposition of Nix being the 1 tool you need (because you need two different Nixes)

2. To motivate abstracting out conventional Unix in Nix so we can then go on to support weird stuff like CloudABI and WASI.

So in both cases, not getting stuff upstreamable is a show-stopper for me. That's why I keep on going on about upstreaming like a broken record :).

(All that said, we'd definitely would want to keep enough branches that we are not actually loosing functionality; and am I quite happe keep on merging a MinGW-only branch into a MSVC too branch so your cool stuff isn't blocked by my upstreaming obsession.)

**Comment by volth** [comment]
If `meson` can have switches like `--with-boost=<path>`, then meson build can be run in MSYS (upstream-legacy way).


And outside MSYS with something like

```bat
for %%gcc in ('nix-build.exe -E "(import <nixpkgs>).mingwPacman.gcc"') do
for %%meson in ('nix-build.exe -E "(import <nixpkgs>).mingwPacman.meson"') do
for %%boost in ('nix-build.exe -E "(import <nixpkgs>).mingwPacman.boost"') do
for %%curl in ('nix-build.exe -E "(import <nixpkgs>).mingwPacman.curl"') do (
  PATH=%%gcc;%PATH%
  %%meson\mingw32\bin\meson.exe ^
  --with-boost=%%boost\mingw32\lib ^
  --with-curl=%%curl\mingw32\lib
)
```
to get MinGW build, and

```bat
for %%stdenvcc in ('nix-build.exe -E "(import <nixpkgs>).stdenv.cc"') do
for %%meson in ('nix-build.exe -E "(import <nixpkgs>).mingwPacman.meson"') do
for %%boost in ('nix-build.exe -E "(import <nixpkgs>).boost"') do
for %%curl in ('nix-build.exe -E "(import <nixpkgs>).curl"') do
  PATH=%%stdenvcc;%PATH%
  %%meson\mingw32\bin\meson.exe ^
  --with-boost=%%boost\lib ^
  --with-curl=%%curl\lib
```
for MSVC

**Comment by Ericson2314** [comment]
Oh meson has tons of Windows support and might already have some boost-specific logic for that too. (It at least has special boost logic for Unix.) But Eelco hasn't seem too interested in checking it out, so I don't want to rely on it in all our branches, hence the autoconf only merge-this-first-please branch. The rest can use Meson/CMake cause yes it's much better.

**Comment by volth** [comment]
I switched `windows` branch to Meson.
There are tons of new `if sys_name.contains('windows')`, but half of them are to cover difference in our versions (`windows` branch is still builds old Nix `2.1.3`), and they will go on version sync

**Comment by Ericson2314** [comment]
Excellent! 

Because `libsodium_dep` and friends are `requred: false` if it not found, I think those deps lists can be unconditional?
